File: CPP/3-Behavioral/Command/Command.cpp
#include <iostream>
#include <vector>

class Order
{
public:
    void buy(int &quantity)
    {
        std::cout << "Buy "
                  << quantity << " Items"
                  << std::endl;
    }
};

class Command
{
public:
    virtual ~Command() {}
    virtual void execute(int &) = 0;
};

class SimpleCommand: public Command
{
    // Define a type to match the function to call
    typedef void(Order:: *Action)(int &);
    Order *_receiver;
    Action _action;

public:
    SimpleCommand(Order *rec, Action act)
    {
        _receiver = rec;
        _action = act;
    }

    /*virtual*/ void execute(int &num)
    {
        // Pass the num argument to the called function
        (_receiver->*_action)(num);
    }
};

class MacroCommand: public Command
{
    std::vector<Command *> _list;

public:
    void add(Command *cmd)
    {
        _list.push_back(cmd);
    }

    /*virtual*/ void execute(int &num)
    {
        for (int i = 0; i < _list.size(); i++)
        {
            _list[i]->execute(num);
        }
    }
};

int main()
{
    Order object;
    Command *commands[2];
    SimpleCommand simpleCommand(&object, &Order::buy);
    commands[0] = &simpleCommand;

    MacroCommand buy_twice;
    buy_twice.add(commands[0]);
    buy_twice.add(commands[0]);
    commands[1] = &buy_twice;

    int num, orderer_index;
    while (true)
    {
        std::cout << "Order: ";
        std::cin >> num;
        std::cout << "Times (1 => 1x 2 => 2x): ";
        std::cin >> orderer_index;
        if (orderer_index == 1 || orderer_index == 2) {
            commands[orderer_index-1]->execute(num);
        }
    }

    return 0;
}

//------------------------------------------------------------------------------
/*
Output:

Enter number selection (0=1x 1=2x): 3 0
   3
Enter number selection (0=1x 1=2x): 3 1
   12
Enter number selection (0=1x 1=2x): 4 0
   8
Enter number selection (0=1x 1=2x): 4 1
   16
*/

File: Python/3-Behavioral/Command/Command.py
#!/usr/bin/env python3

from abc import ABC, abstractmethod


class Order:
    """
    Fake class used with the command pattern
    """
    def buy(self, quantity):
        print("Buy %d Items" % quantity)


class Command(ABC):
    """
    Base abstract class used to define the API for the command
    """
    def __init__(self):
        pass

    @abstractmethod
    def execute(self, val):
        pass


class SimpleCommand(Command):
    def __init__(self, action):
        Command.__init__(self)
        self._action = action

    def execute(self, val):
        print("Running Simple command...")
        self._action(val)


class MacroCommand(Command):
    def __init__(self):
        Command.__init__(self)
        self._commands = []

    def add_command(self, c):
        self._commands.append(c)

    def execute(self, val):
        print("\nRunning macro command...")
        for c in self._commands:
            c.execute(val)


def main():
    order = Order()

    simple_command = SimpleCommand(order.buy)
    buy_twice = MacroCommand()
    buy_twice.add_command(simple_command)
    buy_twice.add_command(simple_command)

    commands = [simple_command, buy_twice]
    for c in commands:
        c.execute(5)


if __name__ == '__main__':
    main()


"""
Output:

Running Simple command...
Buy 5 Items

Running macro command...
Running Simple command...
Buy 5 Items
Running Simple command...
Buy 5 Items
"""
